Revise Mycoportal report fields

JoeCohen opened the pull request:
Revises the `Mycoportal` report fields per communications with MCP
- Updates field/column names
- Updates cell contents
- Adds `coordinateUncertaintyInMeters`

### Suggested Manual Test

- Perform an Observation Search. I suggest creating your own search which doesn't have many hits. 
(I use http://localhost:3000/observations?pattern=sarcosphaera+coronaria+user%3A4468+date%3A2019, which has 3 hits.)
- Click Download Observations
- Select the Mycoportal radio button
- Click Download
- Save the resulting `tsv` in tmp
- Open the saved `tsv` and see if makes sense.
- Run `script/mycoportal_images_only`
- Open the saved `..._images.tsv` and see if it makes sense.

JoeCohen committed:
Change field labels per email from MCP

JoeCohen committed:
Update comment

JoeCohen committed:
change label observationUrl to verbatimAttributes

JoeCohen committed:
Revise value of verbatimAttributes

Change it from the obs url to a formatted link to the obs

JoeCohen committed:
Remove stray temp debugging variable

I had adding this for debugging purposes, forgot to remove it.

JoeCohen committed:
DRY ReportTest mycoportal methods

- Uses a hash to define expected results
- Adds reasonable default values to hash
- Deletes unreadable `expect` array

JoeCohen committed:
RuboCop

JoeCohen committed:
Conform method names to MCP field names

JoeCohen committed:
Constant for Trees/Shrubs for use in associatedTaxa

JoeCohen committed:
Revise host/associated taxa per MCP documentation

- Changes `host` field to `associated_taxa`
- includes observation.notes `host:`  in associated_taxa
- includes observation.notes `Trees/Shrubs:`  in associated_taxa
See https://github.com/BioKIC/symbiota-docs/issues/36#issuecomment-1015733243

JoeCohen committed:
Update expects to comply with revised host/associatedTaxa

JoeCohen committed:
Revise name fields

Amends
- sciname
- scientificNameAuthorship
- identificationRemarks

Deletes
- genus
- specificEpithet

Leaves infraspecificEpithet for the moment. (Not 100% sure if Scott wants it deleted.)

JoeCohen committed:
identificationQualifier for provisionals

JoeCohen committed:
Pass name instead of row to "provisional" methods

JoeCohen committed:
Add Name#provisional?

JoeCohen committed:
Use Name.provisional? instead of Mycoportal.provisional?

JoeCohen committed:
Add id qualifier for names sensu anything

JoeCohen committed:
Add basisOfRecord to report

JoeCohen committed:
Add catalogNumber

JoeCohen committed:
Tweak class and `label` documentation

JoeCohen committed:
Merge branch 'main' into jdc-mycoportal-revised-columns-all

JoeCohen committed:
Merge branch 'main' into jdc-mycoportal-revised-columns-all

JoeCohen committed:
Fix bug in hashed_expect

If Location low/high is nil, expect "", not zero

JoeCohen committed:
continue uncertainy

JoeCohen committed:
test of uncertainty for obs without lat/lon

JoeCohen committed:
Install haversine gem

JoeCohen committed:
Add uncertainty fo obss without lat/lon

JoeCohen committed:
Uncertainty for obs with hidden lat/lon

JoeCohen committed:
Trivial refactor of coordinate_uncertainty

JoeCohen committed:
RuboCopping

JoeCohen committed:
Delete unnecessary fields

JoeCohen committed:
Add author to Coprinus fixture

Need this for testing obs with ID name sensu x

JoeCohen committed:
Fix name_authorship of name sensu whatever

Make it the author of the Name without "sensu"

JoeCohen committed:
refactor identification_qualifier

JoeCohen committed:
Make recordNumber blank unless collection number

JoeCohen committed:
Revise methods which use(d) added columns

- Wrap magic column numbers in explaining methods
- Get `recordedBy` directly from user
- Refactor number
  - rename it to `lowest_collection_number`
  - other refactoring for clarity

JoeCohen committed:
Merge branch 'main' into jdc-mycoportal-revised-columns-all

JoeCohen committed:
Privatize some methods

JoeCohen committed:
Add coverage for DownloadsController

- Covers only previously uncovered line

JoeCohen committed:
Conform order to MCP API output

- Re-orders (arrays, hashes, methods) to conform to the order of the fields in the output of Mycoportal API calls.
This reduces mental gymnastics in comparing our files to MCP output.

JoeCohen committed:
Conform top-level method names to MCP field names

Reduces mental gymnastics

JoeCohen committed:
RuboCop TrailingCommaInHashLiteral

JoeCohen committed:
Delete taxonRank field

MCP automatically generates this from `sciname` and `scientificNameAuthorship`

> <<family, genus, specificEpithet, taxonRank, infraspecificEpithet, scientificNameAuthorship: Do MO supply these or do you derive them from sciname family: Can we supply it for the following types of names (for which it seems to be missing):infrageneric, https://www.mycoportal.org/portal/api/v2/occurrence/8180030“group”, https://www.mycoportal.org/portal/api/v2/occurrence/8179948 provisional, https://www.mycoportal.org/portal/api/v2/occurrence/8180199>>
>
> We just need a species name (sciname) AND authors (scientificNameAuthorship) fields, the rest (e.g., family <according to our thesaurus> and genus etc.) is automatically generated with the upload. We (globally in the Symbiota community) are working toward cleaning up our taxonomica thesaurus and linking it directly to the source data (e.g., via a MycoBank number). We currently don't have a good way of handling provisional names (we added them in the past to the authorship, but this is less than ideal - the community is still in discussion on this).

JoeCohen committed:
Fix authorship and qualifier for qualified names

JoeCohen committed:
Refactor scientific_name_authorship

- Fixes ABC offense
- Other changes to improve clarity

JoeCohen committed:
Add "Sequenced" to remarks if sequenced

JoeCohen committed:
Require haversine gem in report

Fixes uninitialized constant Report::Mycoportal::Haversine when manually testing.
This didn't happen when I running the test suite.

JoeCohen committed:
Account for obss with nil hidden gps

Fixes bug in `coordinate_uncertainty` for Observations with hidden gps but nil lat/lng

JoeCohen committed:
Make verbatimAttributes use webserver url

JoeCohen committed:
Fix typo in verbatimAttributes anchor

JoeCohen committed:
permalink to large image instead of links to all images

JoeCohen committed:
Fix TrailingCommaInHashLiteral

JoeCohen committed:
DRY http_domain

JoeCohen commented on app/classes/report/mycoportal.rb:
```diff
@@ -1,174 +1,341 @@
 # frozen_string_literal: true
 
+require "haversine"
```
Added to fix `uninitialized constant Report::Mycoportal::Haversine `when manually testing.
That's puzzling because tests work fine without this line.

JoeCohen committed:
Comment some methods

JoeCohen commented on Gemfile:
```diff
@@ -157,6 +157,10 @@ gem("prawn-manual_builder")
 
+# calculate the Haversine distance between two points given their lat/lng
+# https://github.com/kristianmandrup/haversine
+gem("haversine")
```
Used to calculate coordinateUncertaintyInMeters

JoeCohen commented on test/controllers/observations/downloads_controller_test.rb:
```diff
@@ -177,6 +177,16 @@ def test_download_observation_index
       assert_no_flash
       assert_response(:success)
+
+      format = "nonexistent"
+      assert_raises("Invalid download type: #{format}") do
+        post(:create,
+             params: {
+               q: query.id.alphabetize,
+               format: format,
+               commit: "Download"
+             })
```
Covers one line in DownloadsController, getting 100% coverage.

JoeCohen commented on test/fixtures/observations.yml:
```diff
@@ -783,6 +783,7 @@ falmouth_2022_obs:
   location: falmouth
   lat: 41.5862
   lng: -70.5976
+  gps_hidden: false
```
A test relies on this value, so I wanted to make it explicit.

JoeCohen commented on test/fixtures/names.yml:
```diff
@@ -51,9 +51,10 @@ coprinus:
-  display_name: '**__Coprinus__**'
-  search_name: Coprinus
-  sort_name: Coprinus
+  author: Pers.
+  display_name: '**__Coprinus__** Pers.'
+  search_name: Coprinus Pers.
+  sort_name: Coprinus  Pers.
```
Needed for a test of an Observation with id "Coprinus sensu lato"

JoeCohen commented on app/classes/report/mycoportal.rb:
```diff
@@ -1,174 +1,348 @@
-    # These plus Column labels comprising
-    # Symbiota Standard Field names which are useful for our purposes
+    # http_domain for links to Observations and Images
+    HTTP_DOMAIN = "https://mushroomobserver.org"
```
Is there a Railsy way of doing this?
(I want a locally-generated report to have links to the website, rather than local links.)

mo-nathan replied:
What's not "Railsy" about that?

JoeCohen replied:
I thought there might be a variable I could access instead of hard-coding it.

JoeCohen committed:
Fix uncertainty when gps hidden && location nil

mo-nathan commented on app/classes/report/mycoportal.rb:
```diff
@@ -1,174 +1,349 @@
+
+    private
+
+    def obs(row)
+      Observation.find(row.obs_id)
+    end
```
Sure.  However, I don't think you really want to find an observation for each row.  That's going to be an N+1 problem which these reports are trying very hard to avoid.  Why can't you use the accessors on the Row object like `row.obs_lat` etc.?

Let me know if you really need to query for the Observation for each row and I can show you how to do that.

JoeCohen replied:
Thanks! I probably will need that, but want to study it more.
I'll get back to you.

JoeCohen committed:
Refactor distance_to_x_corner to take row instead of loc

The first step in trying to get rid of (some) calls to obs(row).

JoeCohen committed:
Delete one call to obs(row)

JoeCohen committed:
Make provisional? a method of row instead of Name

Avoids some calls to obs(row).
(But the method is unavailable to non-Report classes.)

JoeCohen committed:
DRY check whether name is a Group

JoeCohen committed:
Remove test of Name.provisional?

Neglected to do this when moving `provisional?` from Name to Mycoportal

JoeCohen committed:
Calculate location center

Removes another dependency on obs(row).
BUT it's wetter.

JoeCohen committed:
DRY location calculations using Mappable::Box

JoeCohen committed:
Remove scientificNameAuthorship

MCP our value.  It calculates scinetificNameAuthorship based on its own taxonomy.

mo-nathan committed:
Merge branch 'main' into jdc-mycoportal-revised-columns-all

mo-nathan committed:
Remove N+1 queries from MyCoPortal report

JoeCohen committed:
Revise uncertainty for hidden geoloc

- uncertainy = center to farthest corner
- Fixes test failures

JoeCohen committed:
Refactor coordinate_uncertainty

JoeCohen committed:
Cleanup after refactoring uncertainty

- Simplifies private methods used by coordinate_uncertainty

JoeCohen committed:
Merge pull request #2972 from MushroomObserver/njw-mycoportal-query-optimization

Remove N+1 queries from MyCoPortal report

JoeCohen committed:
Delete commented-out code

JoeCohen committed:
Delete now-unused obs(row) method

JoeCohen committed:
Add missing coverage

JoeCohen committed:
Refacotr to tighten report methods

- Return nil instead of ""
- Refactor to reduce number of lines

JoeCohen committed:
Merge branch 'main' into jdc-mycoportal-revised-columns-all

JoeCohen committed:
Include all images instead of only thumbnail

JoeCohen committed:
Draft script to create images-only tsv

Makes a filtered copy of the full Mycoportal report with only the
catalogNumber and imageUrls.

JoeCohen committed:
Merge pull request #2997 from MushroomObserver/mcp-images-only-tsv

MyCoPortal images-only tsv

JoeCohen committed:
Fix typo, make script executable

JoeCohen committed:
Add sample usage comments

JoeCohen committed:
Add report with only catalogNumber and images

Adds a separate spreadsheet to be used to upload just the images to MyCoPortal.

I had tried to do by reading the full report and deleting the unused columns.
But the script which I used threw
```ruby
Illegal quoting in line 119. (CSV::MalformedCSVError)
```

JoeCohen committed:
Add UI for Mycoportal Images report

JoeCohen committed:
RuboCop Fix CyclomaticComplexity offense

JoeCohen committed:
Derive report class from format vs case statement

Gets rid of ugly case statement

JoeCohen committed:
Change internal representation of Darwin report

Fixes `NoMethodError: undefined method `new' for module Report::Darwin`

JoeCohen committed:
Delete imageUrls field from Mycoportal report

- Deletes the field and methods used to calculate the field
- Deletes the data extension which includes that field, and reorders other extended fields accordingly

MyCoPortal doesn't use this field.
Removing it not only slims the spreadsheet, but also eliminates one AR query
That should speed up things up.

JoeCohen committed:
RuboCop delete useless variable

JoeCohen committed:
Delete dateLastModified from Mycoportal report

- Updates tests and expectations
- Removes label and value

Mycoportal does not use this field.

JoeCohen committed:
RuboCop fix alignment offense

JoeCohen committed:
Change Mycoportal report from TSV to CSV
